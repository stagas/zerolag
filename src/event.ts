/* eslint-disable @typescript-eslint/no-explicit-any */
interface EventHandler {
  (a?: any, b?: any, c?: any, d?: any): void
}

export class Event {
  _handlers: Record<string, EventHandler[]> = {}
  silent?: boolean

  _getHandlers(name: string) {
    return (this._handlers[name] = this._handlers[name] || [])
  }
  emit(name: string, a?: any, b?: any, c?: any, d?: any) {
    if (this.silent) return
    const handlers = this._getHandlers(name)
    for (let i = 0; i < handlers.length; i++) {
      handlers[i](a, b, c, d)
    }
  }
  on(name: string, ...newHandlers: EventHandler[]) {
    this._getHandlers(name).push(...newHandlers)
  }
  off(name: string, handler: EventHandler) {
    const handlers = this._getHandlers(name)
    const index = handlers.indexOf(handler)
    if (~index) handlers.splice(index, 1)
  }
  once(name: string, fn: EventHandler) {
    const handlers = this._getHandlers(name)
    const handler = function (a?: any, b?: any, c?: any, d?: any) {
      fn(a, b, c, d)
      handlers.splice(handlers.indexOf(handler), 1)
    }
    handlers.push(handler)
  }
}
