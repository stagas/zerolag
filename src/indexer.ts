interface TextBuffer {
  raw: string
}
export class Indexer {
  buffer: TextBuffer
  constructor(buffer: TextBuffer) {
    this.buffer = buffer
  }
  find(s: string): number[] {
    if (!s) return []
    const offsets = []
    const text = this.buffer.raw
    const len = s.length
    let index = -1
    while (~(index = text.indexOf(s, index + len))) {
      offsets.push(index)
    }
    return offsets
  }
}
