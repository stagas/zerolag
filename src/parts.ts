import { binarySearch } from './util'

function last<T>(array: T[]): T {
  return array[array.length - 1]
}

function remove<T>(array: T[], a: number, b?: number): T[] {
  if (b == null) {
    return array.splice(a)
  } else {
    return array.splice(a, b - a)
  }
}

function insert<T>(target: T[], index: number, array: number[]) {
  const op = array.slice() as [number, number]
  op.unshift(index, 0)
  target.splice.apply(target, op)
}

interface Part extends Array<number> {
  slice(): Part
  concat(n: Part): Part
  startIndex: number
  startOffset: number
}

export interface Find {
  offset: number
  index: number
  local: number
  part: Part
  partIndex: number
}

export class Parts {
  minSize: number
  parts: Part[]
  length: number

  constructor(minSize?: number) {
    minSize = minSize || 5000
    this.minSize = minSize
    this.parts = []
    this.length = 0
  }
  push(item: number) {
    this.append([item])
  }
  append(items: number[]) {
    let part = last(this.parts)

    if (!part) {
      part = [] as never
      part.startIndex = 0
      part.startOffset = 0
      this.parts.push(part)
    } else if (part.length >= this.minSize) {
      const startIndex = part.startIndex + part.length
      const startOffset = items[0]

      part = [] as never
      part.startIndex = startIndex
      part.startOffset = startOffset
      this.parts.push(part)
    }

    part.push.apply(
      part,
      items.map(offset => offset - part.startOffset)
    )

    this.length += items.length
  }
  get(index: number) {
    const part = this.findPartByIndex(index).item
    if (!part) return -1
    return (
      (part[Math.min(part.length - 1, index - part.startIndex)] ?? -1) +
      part.startOffset
    )
  }
  find(offset: number): Find | null {
    const p = this.findPartByOffset(offset)
    if (!p.item) return null

    const part = p.item
    const partIndex = p.index
    const o = this.findOffsetInPart(offset, part)
    return {
      offset: o.item! + part.startOffset,
      index: o.index + part.startIndex,
      local: o.index,
      part: part,
      partIndex: partIndex,
    }
  }
  insert(offset: number, array: number[]) {
    const o = this.find(offset)
    if (!o) {
      return this.append(array)
    }
    if (o.offset > offset) o.local = -1
    const length = array.length
    //TODO: maybe subtract 'offset' instead ????????????????????
    array = array.map(el => (el -= o.part.startOffset))
    insert(o.part, o.local + 1, array)
    this.shiftIndex(o.partIndex + 1, -length)
    this.length += length
  }
  shiftOffset(offset: number, shift: number) {
    const parts = this.parts
    const item = this.find(offset)
    if (!item) return
    if (offset > item.offset) item.local += 1

    let removed = 0
    for (let i = item.local; i < item.part.length; i++) {
      item.part[i] += shift
      if (item.part[i] + item.part.startOffset < offset) {
        removed++
        item.part.splice(i--, 1)
      }
    }
    if (removed) {
      this.shiftIndex(item.partIndex + 1, removed)
      this.length -= removed
    }
    for (let i = item.partIndex + 1; i < parts.length; i++) {
      parts[i].startOffset += shift
      if (parts[i].startOffset < offset) {
        if (last(parts[i]) + parts[i].startOffset < offset) {
          removed = parts[i].length
          this.shiftIndex(i + 1, removed)
          this.length -= removed
          parts.splice(i--, 1)
        } else {
          this.removeBelowOffset(offset, parts[i])
        }
      }
    }
  }
  removeRange(range: [number, number]) {
    const a = this.find(range[0])
    const b = this.find(range[1])
    if (!a || !b) return

    if (a.partIndex === b.partIndex) {
      if (a.offset >= range[1] || a.offset < range[0]) a.local += 1
      if (b.offset >= range[1] || b.offset < range[0]) b.local -= 1
      const shift = remove(a.part, a.local, b.local + 1).length
      this.shiftIndex(a.partIndex + 1, shift)
      this.length -= shift
    } else {
      if (a.offset >= range[1] || a.offset < range[0]) a.local += 1
      if (b.offset >= range[1] || b.offset < range[0]) b.local -= 1
      const shiftA = remove(a.part, a.local).length
      const shiftB = remove(b.part, 0, b.local + 1).length
      if (b.partIndex - a.partIndex > 1) {
        const removed = remove(this.parts, a.partIndex + 1, b.partIndex)
        const shiftBetween = removed.reduce((p, n) => p + n.length, 0)
        b.part.startIndex -= shiftA + shiftBetween
        this.shiftIndex(
          b.partIndex - removed.length + 1,
          shiftA + shiftB + shiftBetween
        )
        this.length -= shiftA + shiftB + shiftBetween
      } else {
        b.part.startIndex -= shiftA
        this.shiftIndex(b.partIndex + 1, shiftA + shiftB)
        this.length -= shiftA + shiftB
      }
    }

    //TODO: this is inefficient as we can calculate the indexes ourselves
    if (!a.part.length) {
      this.parts.splice(this.parts.indexOf(a.part), 1)
    }
    if (!b.part.length) {
      this.parts.splice(this.parts.indexOf(b.part), 1)
    }
  }
  shiftIndex(startIndex: number, shift: number) {
    for (let i = startIndex; i < this.parts.length; i++) {
      this.parts[i].startIndex -= shift
    }
  }
  removeBelowOffset(offset: number, part: Part) {
    const o = this.findOffsetInPart(offset, part)
    const shift = remove(part, 0, o.index).length
    //TODO: this.shiftIndex(o.partIndex + 1, shift) ???????????
    this.shiftIndex(o.index + 1, shift)
    this.length -= shift
  }
  findOffsetInPart(offset: number, part: Part) {
    offset -= part.startOffset
    return binarySearch(part, o => o <= offset)
  }
  findPartByIndex(index: number) {
    return binarySearch(this.parts, s => s.startIndex <= index)
  }
  findPartByOffset(offset: number) {
    return binarySearch(this.parts, s => s.startOffset <= offset)
  }
  toArray() {
    return this.parts.reduce((p, n) => p.concat(n), [] as Part[])
  }
  slice() {
    const parts = new Parts(this.minSize)
    this.parts.forEach((part: Part) => {
      const p = part.slice()
      p.startIndex = part.startIndex
      p.startOffset = part.startOffset
      parts.parts.push(p)
    })
    parts.length = this.length
    return parts
  }
}
