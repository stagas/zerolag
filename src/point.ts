/* eslint-disable @typescript-eslint/no-explicit-any */
export interface PointLike {
  x?: number
  y?: number
  width?: number
  height?: number
}
export interface AreaLike {
  begin: PointLike
  end: PointLike
}
export class Point {
  x = 0
  y = 0

  declare ['/']: Point['div']
  declare ['_/']: Point['floorDiv']
  declare ['o/']: Point['roundDiv']
  declare ['^/']: Point['ceilDiv']
  declare plus: Point['add']
  declare minus: Point['sub']
  declare ['+']: Point['plus']
  declare ['-']: Point['minus']
  declare ['*']: Point['mul']
  declare ['^*']: Point['ceilMul']
  declare ['o*']: Point['roundMul']
  declare ['_*']: Point['floorMul']

  constructor(p?: PointLike) {
    if (p) {
      this.x = p.x!
      this.y = p.y!
    }
  }
  set(p: PointLike) {
    this.x = p.x!
    this.y = p.y!
  }
  isNotZero() {
    return this.x !== 0 || this.y !== 0
  }
  copy() {
    return new Point(this)
  }
  equal(p: PointLike) {
    return this.x === p.x && this.y === p.y
  }
  addRight(x: number) {
    this.x += x
    return this
  }
  abs() {
    return new Point({
      x: Math.abs(this.x),
      y: Math.abs(this.y),
    })
  }
  sign() {
    return new Point({
      x: Math.sign(this.x),
      y: Math.sign(this.y),
    })
  }
  div(p: PointLike) {
    return new Point({
      x: this.x / (p.x || p.width || 0),
      y: this.y / (p.y || p.height || 0),
    })
  }
  floorDiv(p: PointLike) {
    return new Point({
      x: (this.x / (p.x || p.width || 0)) | 0,
      y: (this.y / (p.y || p.height || 0)) | 0,
    })
  }
  roundDiv(p: PointLike) {
    return new Point({
      x: Math.round(this.x / (p.x || p.width || 0)),
      y: Math.round(this.y / (p.y || p.height || 0)),
    })
  }
  ceilDiv(p: PointLike) {
    return new Point({
      x: Math.ceil(this.x / (p.x || p.width || 0)),
      y: Math.ceil(this.y / (p.y || p.height || 0)),
    })
  }
  add(p: PointLike) {
    return new Point({
      x: this.x + (p.x || p.width || 0),
      y: this.y + (p.y || p.height || 0),
    })
  }
  sub(p: PointLike) {
    return new Point({
      x: this.x - (p.x || p.width || 0),
      y: this.y - (p.y || p.height || 0),
    })
  }
  mul(p: PointLike) {
    return new Point({
      x: this.x * (p.x || p.width || 0),
      y: this.y * (p.y || p.height || 0),
    })
  }
  ceilMul(p: PointLike) {
    return new Point({
      x: Math.ceil(this.x * (p.x || p.width || 0)),
      y: Math.ceil(this.y * (p.y || p.height || 0)),
    })
  }
  roundMul(p: PointLike) {
    return new Point({
      x: Math.round(this.x * (p.x || p.width || 0)),
      y: Math.round(this.y * (p.y || p.height || 0)),
    })
  }
  floorMul(p: PointLike) {
    return new Point({
      x: (this.x * (p.x || p.width || 0)) | 0,
      y: (this.y * (p.y || p.height || 0)) | 0,
    })
  }
  lerp(p: PointLike, a: number) {
    return new Point({
      x: this.x + (p.x! - this.x) * a,
      y: this.y + (p.y! - this.y) * a,
    })
  }
  clamp(area: AreaLike) {
    return Point.clamp(area, this)
  }
  toString() {
    return this.x + ',' + this.y
  }
  static sort(a: PointLike, b: PointLike) {
    return a.y! === b.y! ? a.x! - b.x! : a.y! - b.y!
  }
  static gridRound(b: PointLike, a: PointLike) {
    return {
      x: Math.round(a.x! / b.width!),
      y: Math.round(a.y! / b.height!),
    }
  }
  static low(low: PointLike, p: PointLike) {
    return {
      x: Math.max(low.x!, p.x!),
      y: Math.max(low.y!, p.y!),
    }
  }
  static clamp(area: AreaLike & PointLike, p: PointLike) {
    return new Point({
      x: Math.min(
        area.end?.x ?? area.width!,
        Math.max(area.begin?.x ?? 0, p.x!)
      ),
      y: Math.min(
        area.end?.y ?? area.height!,
        Math.max(area.begin?.y ?? 0, p.y!)
      ),
    })
  }
  static offset(b: PointLike, a: PointLike) {
    return { x: a.x! + b.x!, y: a.y! + b.y! }
  }
  static offsetX(x: number, p: PointLike) {
    return { x: p.x! + x, y: p.y! }
  }
  static offsetY(y: number, p: PointLike) {
    return { x: p.x!, y: p.y! + y }
  }
  static toLeftTop(p: PointLike) {
    return {
      left: p.x,
      top: p.y,
    }
  }
}

const proto = Point.prototype as any

proto['/'] = proto.div
proto['_/'] = proto.floorDiv
proto['o/'] = proto.roundDiv
proto['^/'] = proto.ceilDiv
proto['+'] = proto.plus = proto.add
proto['-'] = proto.minus = proto.sub
proto['*'] = proto.mul
proto['^*'] = proto.ceilMul
proto['o*'] = proto.roundMul
proto['_*'] = proto.floorMul
