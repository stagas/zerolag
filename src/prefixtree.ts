// var WORD = /\w+/g;
const WORD = /[a-zA-Z0-9]{1,}/g
// var rank = 0;

export class PrefixTreeNode {
  value: string
  rank: number
  children: Record<string, PrefixTreeNode>
  constructor() {
    this.value = ''
    this.rank = 0
    this.children = {}
  }
  getChildren(): PrefixTreeNode[] {
    const children = Object.keys(this.children).map(key => this.children[key])

    return children.reduce((p, n) => p.concat(n.getChildren()), children)
  }
  collect(key: string) {
    let collection: PrefixTreeNode[] = []
    const node = this.find(key)
    if (node) {
      collection = node
        .getChildren()
        .filter(node => node.value)
        .sort((a, b) => {
          let res = b.rank - a.rank
          if (res === 0) res = b.value.length - a.value.length
          if (res === 0) res = (a.value > b.value) as unknown as number
          return res
        })

      if (node.value) collection.push(node)
    }
    return collection
  }
  find(key: string) {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    let node: PrefixTreeNode = this
    for (const char in key as unknown as string[]) {
      if (key[char] in node.children) {
        node = node.children[key[char]]
      } else {
        return
      }
    }
    return node
  }
  insert(s: string) {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    let node: PrefixTreeNode = this
    let i = 0
    const n = s.length

    while (i < n) {
      if (s[i] in node.children) {
        node = node.children[s[i]]
        i++
      } else {
        break
      }
    }

    while (i < n) {
      node = node.children[s[i]] = node.children[s[i]] || new PrefixTreeNode()
      i++
    }

    node.value = s
    node.rank++
  }
  index(s: string) {
    let word
    while ((word = WORD.exec(s))) {
      this.insert(word[0])
    }
  }
}
