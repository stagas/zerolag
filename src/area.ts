/* eslint-disable @typescript-eslint/no-explicit-any */
import { AreaLike, Point } from './point'

export class Area {
  static offset(b: Area, a: Area) {
    return {
      begin: Point.offset(b.begin, a.begin),
      end: Point.offset(b.end, a.end),
    }
  }

  static offsetX(x: number, a: Area) {
    return {
      begin: Point.offsetX(x, a.begin),
      end: Point.offsetX(x, a.end),
    }
  }

  static offsetY(y: number, a: Area) {
    return {
      begin: Point.offsetY(y, a.begin),
      end: Point.offsetY(y, a.end),
    }
  }

  static sort(a: Area, b: Area) {
    return a.begin.y === b.begin.y
      ? a.begin.x - b.begin.x
      : a.begin.y - b.begin.y
  }

  static toPointSort(a: Area, b: Point) {
    return a.begin.y <= b.y && a.end.y >= b.y
      ? a.begin.y === b.y
        ? a.begin.x - b.x
        : a.end.y === b.y
        ? a.end.x - b.x
        : 0
      : a.begin.y - b.y
  }

  static join(areas: Area[]) {
    const sorted = areas.map(a => a.get()).flatMap(a => [a.begin, a.end]).sort(Point.sort)
    const begin = sorted[0]
    const end = sorted.at(-1)!
    return new Area({ begin, end })
  }

  begin: Point
  end: Point

  constructor(a?: AreaLike) {
    if (a) {
      this.begin = new Point(a.begin)
      this.end = new Point(a.end)
    } else {
      this.begin = new Point()
      this.end = new Point()
    }
  }

  copy() {
    return new Area(this)
  }

  get() {
    const s = [this.begin, this.end].sort(Point.sort)
    return new Area({
      begin: new Point(s[0]),
      end: new Point(s[1]),
    })
  }

  set(area: Area) {
    this.begin.set(area.begin)
    this.end.set(area.end)
  }

  get height() {
    const { begin, end } = this.get()
    return end.y - begin.y
  }

  setLeft(bx: number, ex: number) {
    this.begin.x = bx
    if (ex != null) this.end.x = ex
    return this
  }

  addRight(x: number) {
    this.begin.x += x
    this.end.x += x
    return this
  }

  addBottom(y: number) {
    this.end.y += y
    return this
  }

  shiftByLines(y: number) {
    this.begin.y += y
    this.end.y += y
    return this
  }

  normalizeY() {
    return this.shiftByLines(-this.begin.y)
  }

  greaterThan(a: Area) {
    return this.begin.y === a.end.y
      ? this.begin.x > a.end.x
      : this.begin.y > a.end.y
  }

  greaterThanOrEqual(a: Area) {
    return this.begin.y === a.begin.y
      ? this.begin.x >= a.begin.x
      : this.begin.y > a.begin.y
  }

  lessThan(a: Area) {
    return this.end.y === a.begin.y
      ? this.end.x < a.begin.x
      : this.end.y < a.begin.y
  }

  lessThanOrEqual(a: Area) {
    return this.end.y === a.end.y ? this.end.x <= a.end.x : this.end.y < a.end.y
  }

  isEmpty() {
    return this.begin.equal(this.end)
  }

  inside(a: Area) {
    return (this as any)['>'](a) && (this as any)['<'](a)
  }

  outside(a: Area) {
    return (this as any)['<'](a) || (this as any)['>'](a)
  }

  insideEqual(a: Area) {
    return (this as any)['>='](a) && (this as any)['<='](a)
  }

  outsideEqual(a: Area) {
    return (this as any)['<='](a) || (this as any)['>='](a)
  }

  equal(a: Area) {
    return (
      this.begin.x === a.begin.x
      && this.begin.y === a.begin.y
      && this.end.x === a.end.x
      && this.end.y === a.end.y
    )
  }

  beginLineEqual(a: Area) {
    return this.begin.y === a.begin.y
  }

  endLineEqual(a: Area) {
    return this.end.y === a.end.y
  }

  linesEqual(a: Area) {
    return (this as any)['|='](a) && (this as any)['=|'](a)
  }

  sameLine(a: Area) {
    return this.begin.y === this.end.y && this.begin.y === a.begin.y
  }

  shortenByX(x: number) {
    return new Area({
      begin: {
        x: this.begin.x + x,
        y: this.begin.y,
      },
      end: {
        x: this.end.x - x,
        y: this.end.y,
      },
    })
  }

  widenByX(x: number) {
    return new Area({
      begin: {
        x: this.begin.x - x,
        y: this.begin.y,
      },
      end: {
        x: this.end.x + x,
        y: this.end.y,
      },
    })
  }

  toString() {
    const area = this.get()
    return '' + area.begin + '|' + area.end
  }
}

const proto = Area.prototype as any
proto['>'] = proto.greaterThan
proto['>='] = proto.greaterThanOrEqual
proto['<'] = proto.lessThan
proto['<='] = proto.lessThanOrEqual
proto['><'] = proto.inside
proto['<>'] = proto.outside
proto['>=<'] = proto.insideEqual
proto['<=>'] = proto.outsideEqual
proto['==='] = proto.equal
proto['|='] = proto.beginLineEqual
proto['=|'] = proto.endLineEqual
proto['|=|'] = proto.linesEqual
proto['=|='] = proto.sameLine
proto['-x-'] = proto.shortenByX
proto['+x+'] = proto.widenByX
