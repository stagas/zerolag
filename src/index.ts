import { Area } from './area'
import { Point } from './point'
import { Event } from './event'
import { Regexp } from './regexp'

import { SkipString } from './skipstring'
import { PrefixTreeNode } from './prefixtree'
import { Segments } from './segments'
import { Indexer } from './indexer'
import { Tokens } from './tokens'
import { Parts } from './parts'
// import { Syntax } from './syntax'

const EOL = /\r\n|\r|\n/g
const NEWLINE = /\n/g
const WORDS = Regexp.create(['tokens'], 'g')

// const SEGMENT: Record<string, string> = {
//   comment: '/*',
//   string: '`',
// }

interface BufferPoint extends Point {
  offset: number
  line: Line
  point: BufferPoint
}

class Line {
  offsetRange: Range = [-1, -1]
  offset = 0
  length = 0
  point = new Point()
}

function normalizeEOL(s: string) {
  return s.replace(EOL, '\n')
}

type Range = [number, number]

type LogEntry = ['insert' | 'remove', Range, string]

export class Buffer extends Event {
  log: LogEntry[]

  raw = ''

  indexer: Indexer
  segments: Segments

  text!: SkipString
  tokens!: Tokens<Parts>
  prefix!: PrefixTreeNode

  constructor() {
    super()
    this.log = []
    // this.syntax = new Syntax;
    this.indexer = new Indexer(this)
    this.segments = new Segments(this as any)
    this.setText('')
  }
  updateRaw() {
    this.raw = this.text.toString()
  }
  copy() {
    this.updateRaw()
    const buffer = new Buffer()
    buffer.replace(this)
    return buffer
  }
  replace(data: Buffer) {
    this.raw = data.raw
    this.text.set(this.raw)
    this.tokens = data.tokens.copy()
    this.segments.clearCache()
  }
  setText(text: string) {
    text = normalizeEOL(text)

    this.raw = text //this.syntax.highlight(text);

    // this.syntax.tab = ~this.raw.indexOf('\t') ? '\t' : ' ';
    this.text = new SkipString()
    this.text.set(this.raw)

    this.tokens = new Tokens(() => new Parts())
    this.tokens.index(this.raw)
    this.tokens.on('change segments', this.emit.bind(this, 'change segments'))

    this.prefix = new PrefixTreeNode()
    this.prefix.index(this.raw)

    this.emit('set')
  }
  insertTextAtPoint(p: BufferPoint, text: string, noLog: boolean) {
    if (!noLog) this.emit('before update')

    text = normalizeEOL(text)

    const length = text.length
    const point = this.getPoint(p)
    const shift = (text.match(NEWLINE) || []).length
    const range: Range = [point.y, point.y + shift]
    const offsetRange = this.getLineRangeOffsets(range)

    const before = this.getOffsetRangeText(offsetRange)
    this.text.insert(point.offset, text)
    offsetRange[1] += text.length
    const after = this.getOffsetRangeText(offsetRange)
    this.prefix.index(after)
    this.tokens.update(offsetRange, after, length)
    this.segments.clearCache(offsetRange[0])
    if (!noLog)
      this.appendLog('insert', [point.offset, point.offset + text.length], text)

    if (!noLog) this.emit('update', range, shift, before, after)

    return text.length
  }
  removeOffsetRange(o: Range, noLog?: boolean) {
    if (!noLog) this.emit('before update')

    const a = this.getOffsetPoint(o[0])
    const b = this.getOffsetPoint(o[1])
    const length = o[0] - o[1]
    const range: Range = [a.y, b.y]
    const shift = a.y - b.y

    const offsetRange = this.getLineRangeOffsets(range)
    const before = this.getOffsetRangeText(offsetRange)
    const text = this.text.getRange(o)
    this.text.remove(o)
    offsetRange[1] += length
    const after = this.getOffsetRangeText(offsetRange)
    this.prefix.index(after)
    this.tokens.update(offsetRange, after, length)
    this.segments.clearCache(offsetRange[0])
    if (!noLog) this.appendLog('remove', o, text)

    if (!noLog) this.emit('update', range, shift, before, after)
  }
  appendLog(type: 'insert' | 'remove', offsets: Range, text: string) {
    if (type === 'insert') {
      const lastLog = this.log[this.log.length - 1]
      if (lastLog && lastLog[0] === 'insert' && lastLog[1][1] === offsets[0]) {
        lastLog[1][1] += text.length
        lastLog[2] += text
      } else {
        this.log.push(['insert', offsets, text])
      }
    } else if (type === 'remove') {
      const lastLog = this.log[this.log.length - 1]
      if (lastLog && lastLog[0] === 'remove' && lastLog[1][0] === offsets[1]) {
        lastLog[1][0] -= text.length
        lastLog[2] = text + lastLog[2]
      } else {
        this.log.push(['remove', offsets, text])
      }
    }
  }
  removeArea(area: Area) {
    const offsets = this.getAreaOffsetRange(area)
    return this.removeOffsetRange(offsets)
  }
  removeCharAtPoint(p: Point) {
    const point = this.getPoint(p)
    const offsetRange: Range = [point.offset, point.offset + 1]
    return this.removeOffsetRange(offsetRange)
  }
  // get(range) {
  //   let code = this.getLineRangeText(range)

  //   // calculate indent for `code`
  //   //TODO: move to method
  //   const last = code.slice(code.lastIndexOf('\n'))
  //   const AnyChar = /\S/g
  //   let y = range[1]
  //   let match = AnyChar.exec(last)
  //   while (!match && y < this.loc()) {
  //     const after = this.getLineText(++y)
  //     AnyChar.lastIndex = 0
  //     match = AnyChar.exec(after)
  //   }
  //   let indent = 0
  //   if (match) indent = match.index
  //   const indentText = '\n' + new Array(indent + 1).join(this.syntax.tab)

  //   const segment = this.segments.get(range[0])
  //   if (segment) {
  //     code = SEGMENT[segment] + '\uffba\n' + code + indentText + '\uffbe*/`'
  //     code = this.syntax.highlight(code)
  //     code =
  //       '<' +
  //       segment[0] +
  //       '>' +
  //       code.substring(code.indexOf('\uffba') + 2, code.lastIndexOf('\uffbe'))
  //   } else {
  //     code = this.syntax.highlight(code + indentText + '\uffbe*/`')
  //     code = code.substring(0, code.lastIndexOf('\uffbe'))
  //   }
  //   return code
  // }
  getLine(y: number) {
    const line = new Line()
    const loc = this.loc()
    line.offsetRange = this.getLineRangeOffsets([y, y])
    line.offset = line.offsetRange[0]
    line.length = line.offsetRange[1] - line.offsetRange[0] - (y < loc ? 1 : 0)
    line.point.set({ x: 0, y: y >= loc ? loc : y })
    return line
  }
  getPoint(p: Point) {
    const line = this.getLine(p.y)
    const point = new Point({
      x: Math.min(line.length, p.x!),
      y: line.point.y,
    }) as Partial<BufferPoint>
    point.offset = line.offset + point.x!
    point.point = point as BufferPoint
    point.line = line
    return point as BufferPoint
  }
  getLineRangeText(range: [number, number]) {
    const offsets = this.getLineRangeOffsets(range)
    const text = this.text.getRange(offsets)
    return text
  }
  getLineRangeOffsets(range: [number, number]): [number, number] {
    const a = this.getLineOffset(range[0])
    const b =
      range[1] >= this.loc()
        ? this.text.length
        : this.getLineOffset(range[1] + 1)
    const offsets: [number, number] = [a, b]
    return offsets
  }
  getOffsetRangeText(offsetRange: [number, number]) {
    const text = this.text.getRange(offsetRange)
    return text
  }
  getOffsetPoint(offset: number) {
    const token = this.tokens.getByOffset('lines', offset - 0.5)
    if (!token) return new Point()
    return new Point({
      x:
        offset -
        (offset > token.offset
          ? token.offset + (!!token.part.length as unknown as number)
          : 0),
      y: Math.min(
        this.loc(),
        token.index - ((token.offset + 1 > offset) as unknown as number) + 1
      ),
    })
  }
  charAt(offset: number) {
    const char = this.text.getRange([offset, offset + 1])
    return char
  }
  // getOffsetLineText(offset) {
  //   return {
  //     line: line,
  //     text: text,
  //   }
  // }
  getLineLength(line: number) {
    return this.getLine(line).length
  }
  getLineText(y: number) {
    const text = this.getLineRangeText([y, y])
    return text
  }
  getAreaText(area: Area) {
    const offsets = this.getAreaOffsetRange(area)
    const text = this.text.getRange(offsets)
    return text
  }
  wordAreaAtPoint(p: BufferPoint, inclusive?: boolean) {
    const point = this.getPoint(p)
    const text = this.text.getRange(point.line.offsetRange)
    const words = Regexp.parse(text, WORDS)

    if (words.length === 1) {
      const area = new Area({
        begin: { x: 0, y: point.y },
        end: { x: point.line.length, y: point.y },
      })

      return area
    }

    let lastIndex = 0
    let word = []
    let end = text.length

    for (let i = 0; i < words.length; i++) {
      word = words[i]
      if (word.index > point.x - (!!inclusive as unknown as number)) {
        end = word.index
        break
      }
      lastIndex = word.index
    }

    const area = new Area({
      begin: { x: lastIndex, y: point.y },
      end: { x: end, y: point.y },
    })

    return area
  }
  moveAreaByLines(dy: number, area: Area) {
    if (area.begin.y + dy < 0 || area.end.y + dy > this.loc()) return false

    let x = 0
    let y = area.begin.y + dy

    let swap_a = false
    let swap_b = false

    area.end.x = area.begin.x = 0
    area.end.y = area.end.y + 1

    if (dy > 0 && area.end.y === this.loc()) {
      if (area.begin.y === 0) {
        area.begin.x = 0
        area.end.x = 0
        x = Infinity
        swap_b = true
      } else {
        area.end.y = this.loc()
        y = area.begin.y + dy
        x = Infinity
        swap_b = true
      }
    } else if (dy < 0 && area.end.y > this.loc() && y > 0) {
      area.begin.y = y
      area.begin.x = this.getLineLength(area.begin.y)
      y = area.begin.y - 1
      x = Infinity
    } else if (dy < 0 && y === 0 && area.end.y > this.loc()) {
      area.begin.y -= 1
      area.begin.x = this.getLineLength(area.begin.y)
      swap_a = true
    }

    const offsets = this.getAreaOffsetRange(area)
    let text = this.text.getRange(offsets)

    if (swap_a) {
      text = text.slice(1) + text[0]
    }
    if (swap_b) {
      text = text.slice(-1) + text.slice(0, -1)
    }

    this.remove(offsets)
    this.insert({ x, y }, text)

    return true
  }
  getAreaOffsetRange(area: Area): Range {
    const begin = this.getPoint(area.begin)
    const end = this.getPoint(area.end)
    const range: Range = [
      Math.max(0, begin.offset),
      end.y < area.end.y ? end.line.offsetRange[1] : end.offset,
    ]
    return range
  }
  // getOffsetLine(offset) {
  //   return line
  // }
  getLineOffset(y: number) {
    const offset =
      y < 0 ? -1 : y === 0 ? 0 : this.tokens.getByIndex('lines', y - 1) + 1
    return offset
  }
  getLongestLine() {
    return this.getLongestLineLength(true)
  }
  getLongestLineLength(withLineNumber?: boolean) {
    // TODO: this should be part of the 'Parts' class
    // so lookup becomes O(1), currently lookup is O(n)
    let max = this.getLineLength(this.loc()) + 1,
      y = this.loc(),
      diff = 0,
      prev = -1,
      curr = 0
    const parts = this.tokens.getCollection('lines').parts
    for (let i = 0, cy = 0; i < parts.length; i++) {
      const part = parts[i]
      for (let j = 0; j < part.length; j++) {
        cy++
        curr = part[j]
        diff = curr - prev
        prev = curr
        if (diff > max) {
          max = diff
          y = cy
        }
      }
    }
    if (withLineNumber)
      return {
        length: max - 1,
        lineNumber: Math.max(0, y - 1),
      }
    return max - 1 // minus the newline char
  }
  loc() {
    return this.tokens.getCollection('lines').length
  }
  toString() {
    return this.text.toString()
  }

  insert: any
  remove: any
}

// Buffer.prototype.__proto__ = Event.prototype;

Buffer.prototype.insert = Buffer.prototype.insertTextAtPoint
// ;

Buffer.prototype.remove = Buffer.prototype.removeOffsetRange
// ;
