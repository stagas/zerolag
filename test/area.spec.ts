// @env browser
import { Area } from '../src/area'

describe('Area', () => {
  it('join', () => {
    const a = new Area({
      begin: { x: 5, y: 6 },
      end: { x: 2, y: 10 },
    })

    const b = new Area({
      begin: { x: 5, y: 16 },
      end: { x: 2, y: 20 },
    })

    const area = Area.join([a, b])
    expect(area).toMatchSnapshot()
  })
})
