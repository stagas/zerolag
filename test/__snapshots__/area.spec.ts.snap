
exports[`Area join 1`] = `
Area {
  "begin": Point {
    "x": 5,
    "y": 6,
  },
  "end": Point {
    "x": 2,
    "y": 20,
  },
}
`;
